added accesstype check before adding read operations

guj opened the pull request:
this is to avoid have read operations showing up in (multiple) write flushes
for the same iteration

guj committed:
added accesstype check before adding read operations
this is to avoid have read operations showing up in (multiple) write flushes
for the same iteration

guj committed:
fixed tabs

guj committed:
fixed end of line

ax3l committed:
ADIOS1: Activate Skip Iterations Test

Demonstrate an ADIOS1 backend bug seen in WarpX.

ax3l commented:
I added a demonstrator of the bug that we are hunting in #680.

According to the failing tests, it is possible that some attributes are not written after the first `.flush()` or that their values are not read in after re-opening a series with a new `Series` object.

ax3l commented on src/Iteration.cpp:
```diff
@@ -118,6 +118,13 @@ Iteration::flushFileBased(std::string const& filename, uint64_t i)
         IOHandler->enqueue(IOTask(this, pCreate));
     } else
     {
+      if (IOHandler->accessTypeFrontend == AccessType::CREATE)
```
if we can somehow determine that the "old" series has been read at least once, we can also try:
```suggestion
      if (IOHandler->accessTypeFrontend == AccessType::CREATE || IOHandler->accessTypeFrontend == AccessType::READ_WRITE)
```

I would generally be more comfortable if we can place this backend-specific work-around in the ADIOS1IOHandlerImpl file(s) instead of here, we could e.g. return the `OPEN_FILE` call that is triggered here.

(And probably have to add a work-around (skip-if-not-open) to the `OPEN_PATH`, too.)

guj replied:
I think read_write should be allowed to read. That's why I did not include it. 

On the other hand, adios does have have both read/write mode

ax3l replied:
you could add
```suggestion
      if (IOHandler->accessTypeFrontend == AccessType::CREATE && IOHandler->backend() == "MPI_ADIOS1")
```
as a really directed frontend hack (for a backend problem).

ax3l replied:
As discussed on Slack, the most general solution to this problem is likely to open files directly in the backend instead of delaying their open.
That means `createFile` should open files already instead of waiting for `openFile`, `writeDataset`, an `writeAttribute`.
That would also reduce the massive open-close calls frequency.

guj replied:
OK added

guj committed:
set openPMDextension=1 for all iterations

guj committed:
end of line fix

guj committed:
added a helper function

guj committed:
check subgroup exists before creating new one

guj committed:
adios1 hangs when there are many files open
so close old ones before opening new one

guj committed:
fixed tabs

guj committed:
end-of-line fix

guj committed:
removed unused line

guj committed:
coment out unused line

guj committed:
change is limited to ADIOS

guj committed:
tab fix

guj committed:
Serial  ADIOS needs to included too

guj committed:
cleaned up the ADIOS1Impl destructor

guj committed:
removed the workaround and cleaned up code

guj committed:
put the fix for ADIOS1's Parallel bug

guj committed:
fixed tabs

guj committed:
fixed white space

guj committed:
unused int commented out

guj commented:
Found the cause of the problem.  Not a workaround anymore. 
Need help with travis,   times out for  Appleclang11.0

guj committed:
Merge branch 'dev' of https://github.com/openPMD/openPMD-api into adios1-fix-open-in-write

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -722,7 +722,7 @@ void fileBased_write_test(const std::string & backend)
         }
 
-        o.setOpenPMDextension(1);
+        //o.setOpenPMDextension(1); // this will make iteration 1&2 have extension 0.
```
With the root cause found, can we activate this "late" Series-global attribute again, please? The comment behind this was only true for ADIOS1, according to my tests.
If this is only necessary for ADIOS1, let's add another `o.backendName() == "ADIOS1"` if branch here.

guj replied:
ok, found a solution.

ax3l commented on src/IO/ADIOS/CommonADIOS1IOHandler.cpp:
```diff
@@ -803,17 +809,11 @@ CommonADIOS1IOHandlerImpl::writeAttribute(Writable* writable,
-        m_openWriteFileHandles[res->second] = fd;
-    } else
-        fd = m_openWriteFileHandles.at(res->second);
+    //int64_t fd = GetFileHandle(writable);
```
Let's remove the dead code there:
```suggestion
```

ax3l commented on src/IO/ADIOS/CommonADIOS1IOHandler.cpp:
```diff
@@ -749,24 +746,33 @@ CommonADIOS1IOHandlerImpl::deleteAttribute(Writable*,
+                                  Parameter< Operation::WRITE_DATASET > const& parameters)
+{
+    if( m_handler->accessTypeBackend == AccessType::READ_ONLY )
+        throw std::runtime_error("[ADIOS1] Writing into a dataset in a file opened as read only is not possible.");
```
Typo?
```suggestion
        throw std::runtime_error("[ADIOS1] Writing into a dataset in a file opened as read-only is not possible.");
```

ax3l commented on src/IO/ADIOS/CommonADIOS1IOHandler.cpp:
```diff
@@ -372,9 +372,11 @@ CommonADIOS1IOHandlerImpl::createFile(Writable* writable,
         /* defer actually opening the file handle until the first Operation::WRITE_DATASET occurs */
         m_existsOnDisk[m_filePaths[writable]] = false;
+
+        /*int64_t fd = */GetFileHandle(writable);
```
```suggestion
        GetFileHandle(writable);
```

ax3l commented on include/openPMD/IO/ADIOS/ParallelADIOS1IOHandlerImpl.hpp:
```diff
@@ -94,6 +94,7 @@ namespace openPMD
         std::unordered_map< std::shared_ptr< std::string >, ADIOS_FILE* > m_openReadFileHandles;
         std::unordered_map< ADIOS_FILE*, std::vector< ADIOS_SELECTION* > > m_scheduledReads;
         std::unordered_map< int64_t, std::unordered_map< std::string, Attribute > > m_attributeWrites;
+        int64_t GetFileHandle(Writable*);
```
Can you please add a doxygen string that describes the side-effects and return-value of this function?
We call this in several places without handling the return value, which might lead to confusion later on.

guj replied:
added

ax3l commented on include/openPMD/IO/ADIOS/ADIOS1IOHandlerImpl.hpp:
```diff
@@ -93,6 +93,7 @@ namespace openPMD
         std::unordered_map< std::shared_ptr< std::string >, ADIOS_FILE* > m_openReadFileHandles;
         std::unordered_map< ADIOS_FILE*, std::vector< ADIOS_SELECTION* > > m_scheduledReads;
         std::unordered_map< int64_t, std::unordered_map< std::string, Attribute > > m_attributeWrites;
+        int64_t GetFileHandle(Writable*);
```
Can you please add a doxygen string that describes the side-effects and return-value of this function?
We call this in several places without handling the return value, which might lead to confusion later on.

guj replied:
added

ax3l commented on src/IO/ADIOS/ParallelADIOS1IOHandler.cpp:
```diff
@@ -58,19 +58,26 @@ ParallelADIOS1IOHandlerImpl::~ParallelADIOS1IOHandlerImpl()
-            close(f.second);
+            allFiles[*(f.first)] = f.second;
+
+    for (auto p :  allFiles)
```
```suggestion
    for (auto& p :  allFiles)
```

ax3l commented on src/IO/ADIOS/ParallelADIOS1IOHandler.cpp:
```diff
@@ -58,19 +58,26 @@ ParallelADIOS1IOHandlerImpl::~ParallelADIOS1IOHandlerImpl()
+    // which  caused trouble with  close(), which is collective
+    // so I just sort by file name to force  all processors  close
+    // all the fids in the same order
+    std::map< std::string, int64_t> allFiles;
```
The next three lines can probably be [shortened to](https://en.cppreference.com/w/cpp/container/map/map):
```suggestion
    std::map< std::string, int64_t> allFiles(
        m_openWriteFileHandles.begin(), m_openWriteFileHandles.end() );
```

guj replied:
I would prefer to stay put.

ax3l commented on src/IO/ADIOS/ParallelADIOS1IOHandler.cpp:
```diff
@@ -58,19 +58,26 @@ ParallelADIOS1IOHandlerImpl::~ParallelADIOS1IOHandlerImpl()
+
+    for (auto p :  allFiles)
+      {
+        auto fid =  p.second;
```
```suggestion
        auto const fid = p.second;
```

ax3l commented on src/IO/HDF5/HDF5IOHandler.cpp:
```diff
@@ -162,6 +162,10 @@ HDF5IOHandlerImpl::createPath(Writable* writable,
         groups.push(node_id);
         for( std::string const& folder : auxiliary::split(path, "/", false) )
         {
+            htri_t found = H5Lexists(groups.top(), folder.c_str(), H5P_DEFAULT);
```
const-correctness and comment
```suggestion
            // avoid creation of paths that already exist
            htri_t const found = H5Lexists(groups.top(), folder.c_str(), H5P_DEFAULT);
```

ax3l commented on src/Iteration.cpp:
```diff
@@ -118,6 +118,14 @@ Iteration::flushFileBased(std::string const& filename, uint64_t i)
         IOHandler->enqueue(IOTask(this, pCreate));
     } else
     {
+      if ((IOHandler->accessTypeFrontend == AccessType::CREATE)
```
is this work-around still needed?

guj replied:
yes

guj committed:
Update src/IO/HDF5/HDF5IOHandler.cpp

Co-Authored-By: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update src/IO/ADIOS/ParallelADIOS1IOHandler.cpp

Co-Authored-By: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update src/IO/ADIOS/ParallelADIOS1IOHandler.cpp

Co-Authored-By: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update src/IO/ADIOS/CommonADIOS1IOHandler.cpp

Co-Authored-By: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update src/IO/ADIOS/CommonADIOS1IOHandler.cpp

Co-Authored-By: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update src/IO/ADIOS/CommonADIOS1IOHandler.cpp

Co-Authored-By: Axel Huebl <axel.huebl@plasma.ninja>

ax3l commented on src/IO/ADIOS/ParallelADIOS1IOHandler.cpp:
```diff
@@ -58,19 +58,26 @@ ParallelADIOS1IOHandlerImpl::~ParallelADIOS1IOHandlerImpl()
-            close(f.second);
+            allFiles[*(f.first)] = f.second;
+
+    for (auto& p :  allFiles)
```
```suggestion
    for (auto const & p : allFiles)
```

guj replied:
changed

ax3l commented on src/IO/ADIOS/CommonADIOS1IOHandler.cpp:
```diff
@@ -749,24 +746,33 @@ CommonADIOS1IOHandlerImpl::deleteAttribute(Writable*,
     {
-        fd = open_write(writable);
-        m_openWriteFileHandles[res->second] = fd;
+      std::string  name  = *(res->second);
```
```suggestion
      std::string const name = *(res->second);
```

guj committed:
added doxygen to GetFileHandle

guj committed:
when an iteration has been flushed already before  the series flush is called,
the open_file task is created and handled in  adios1 to  make series.m_writer to  point  to the iteration's writer
so that later calls of series.flushAttributes() can work on that iteration
when series is calling flush file based.

guj committed:
back to original (before  this pull request)

guj committed:
minor change on style

guj committed:
tab fix

guj committed:
fixed end of line

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -725,7 +724,7 @@ void fileBased_write_test(const std::string & backend)
         o.iterations[3].setTime(static_cast< double >(3));
         o.iterations[4].setTime(static_cast< double >(4));
-        o.flush();
+        //o.flush();
```
removing that is confusing because it is a late flush. flushing here should be very similar to the destructor which is called next and flushes, too.

what I thought we chatted about was to set for ADIOS the `setOpenPMDextension` before the first iteration's flush occurs, because it seams to set it at that point and cannot be overwritten later on?

guj replied:
Oh I thought I explained that the attributes in adios1  are written when the series is deleted, so it is not really an overwritten issue. 
I can put this flush back. Thought it is not needed. Is there a reason to have it there?

guj replied:
OK I looked in the history flush was there. I put it back. Sorry

ax3l replied:
Nothing to be sorry about, thanks a lot :)

guj committed:
put back the flush that I commented out from original code

guj committed:
Fix Formatting

ax3l committed:
Changelog & Junmin to Co-Authors

Thank you for your contributions!

ax3l commented:
This fixes #680, thank you so much!

ax3l commented on .rodare.json:
```diff
@@ -14,6 +14,12 @@
+            "affiliation": "Lawrence Berkeley National Laboratory",
+            "name": "Gu, Junmin",
+            "orcid": "0000-0002-1521-8534",
+            "type": "Other"
```
fixed in `dev` via 4ca1c9c
